<a href="https://colab.research.google.com/github/bitprj/DigitalHistory/blob/Narae/Week3-Introduction-to-Open-Data-Importing-Data-and-Basic-Data-Wrangling/Week3_Introduction_to_Data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Manipulation with Pandas




## Table of Contents
- Why, Where and How we use Pandas
- What we will be learning today
  - About the dataset
  - Goals
- Connecting to Google Drive
- Importing libraries and unpacking a file
- Loading data
- Getting info about the dataset
  - Now Try This
- Removing NaN (None) values
  - Now Try This
- Selecting subsets of data
  - Now Try This
- Filtering dataset based on criteria
  - Now Try This
- Aggregation functions
  - Now Try This

## Why, Where and How we use Pandas

This week, we will cover the basic data manipulation using Pandas.
1. Pandas is an open source data analysis and manipulation tool and it is widely used both in academia and industry.
2. It is built on top of the Python programming. 
3. It offers data structures and operations for manipulating numerical tables and time series.

Pandas provides three data structures: Series, DataFrame, and Panel. 
1. A Series is 1-dimensional labelled array and it can hold data of **any type** (integer, string, float, python objects, etc.). Its labels are called an index.
2. A DataFrame is 2-dimensional labelled data structure with both rows and columns.
3. A panel is 3-dimensional. 

This week, we will focus on DataFrame and we will learn Series in later weeks. We will not cover Panel in this semester, as it's not used as often as two other data structures.


Since we've covered the fundamentals of Python, it will be fairly easy to pick up Pandas.

## What we will be learning today

### About the dataset:
To begin with Pandas and dataframe, we will use titanic dataset. It has all of the passengers' information, such as name, age, gender, ticket price, and most importantly whether they survived or not.

As each person has its unique PassengerID, each row is a unique entity / passenger.



### Goals:
Using data manipulation and analysis, we will learn how to get a quick overview of the dataset, how to select and filter specific rows / columns. We At the end of the tutorial, we should be able to analyze the survival rates.

## Connecting to Your Google Drive

We will start by connecting google drive into google colab

In [1]:
from google.colab import drive

drive.mount('/content/gdrive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/gdrive


### Change Directory

Now, the google drive is connected to our colab notebook.
Let's access the dataset located in Digital History/Week3.

```cd``` is a command used to change current working directory. Simply, ```cd``` means 'Change Directory.'

By running the code below, we change our current directory.


In [4]:
cd "/content/gdrive/My Drive/DigitalHistory/Week3-Introduction-to-Open-Data-Importing-Data-and-Basic-Data-Wrangling/data"

/content/gdrive/.shortcut-targets-by-id/1m-IVNIRZmHM3YwHOGFHd_tUQuI298irS/DigitalHistory/Week3-Introduction-to-Open-Data-Importing-Data-and-Basic-Data-Wrangling/data


### List files
```ls``` is a command used to list all files in the current directory.

After we run it, we will see a 'titanic-dataset.csv' file.

In [5]:
ls

acs2017_county_data.csv  __MACOSX/  titanic-dataset.csv  titanic-dataset.zip


'titanic-dataset.csv' is the file we are going to use for step-by-step learning session.

### Import Library

In order to read / load a file, we will need to import Pandas.

It's a convention to use ``` import pandas as pd``` when importing Pandas library.

In [6]:
import pandas as pd

Once we've imported Pandas, we can use ``` pd``` to call any functions in Pandas.



### Load file

We will use ```read_csv``` function as we will read a csv file.

Since we are working with only one dataset, we will just call dataframe as df. 

But, if we are working with lots of dataframes, it's better to give a meaningful name (ex: titanic_data, passenger_info, etc.)

In [8]:
df = pd.read_csv('titanic-dataset.csv')
df

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0              1         0       3  ...   7.2500   NaN         S
1              2         1       1  ...  71.2833   C85         C
2              3         1       3  ...   7.9250   NaN         S
3              4         1       1  ...  53.1000  C123         S
4              5         0       3  ...   8.0500   NaN         S
..           ...       ...     ...  ...      ...   ...       ...
886          887         0       2  ...  13.0000   NaN         S
887          888         1       1  ...  30.0000   B42         S
888          889         0       3  ...  23.4500   NaN         S
889          890         1       1  ...  30.0000  C148         C
890          891         0       3  ...   7.7500   NaN         Q

[891 rows x 12 columns]

Now, the dataset is loaded as a dataframe 'df'

## Basic info about the dataset

Now, let's get the basic information about the dataframe.
- head()
- describe()
- info()

### head()
head() function is useful to see the dataset at a quick glance as it returns first n rows.

Let's check what columns this file has by calling 'head()' function.

By default, the head() returns the first 5 rows.

In [9]:
df.head()

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S

[5 rows x 12 columns]

You can specify the number of rows to display by calling df.head(number)

In [ ]:
df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

### info()

Now, we know what's in the dataset and what it looks like.

This function is useful as this returns all of the **column names** and **its types** as well as **Non-Null** counts. 

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


We can tell that "Age" and "Cabin" have lots of missing values. 

If we take a closer look at dtypes in the second to last row, there are three dtypes: float64, int64, and object.

We have covered float64 and int64 last week in Python, but what is an object?

- int64: integer numbers
- float64: floating point numbers
- object: string or mixed numeric and non-numeric values.

That's why the dtype of "Name", "Sex", "Embarked" is object, as it is a string.

"Ticket" and "Cabin" are object as they are in a format of numbers or string+numbers (Ex: A/5 21171, C85)



### describe()

describe() is used to view summary statistics of numeric columns. This helps us to have a general idea of the dataset.

In [ ]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

### shape
To see the size of the dataset, we can use shape function, which returns the number of rows and columns in a format of (# rows, # columns)

This dataset has 891 rows (entities) and 12 columns.

In [ ]:
df.shape

(891, 12)

## Remove NaN values

Often times, when we work with large datasets, we will encounter cases where there are lots of missing elements (NaN / null) in the dataset.

Removing NaN values will allow us to drop the rows and to work with clean datasets.


---


Let's remove the rows that do not provide a meaningful information.

When we know a "unique key" of the dataset (PassengerID in this dataset), we can check whether all elements have PassengerID. If any of the rows are missing PassengerID, then we can drop that entity.

* df.dropna(): drop the rows where at least one of the elements is missing.
* df.dropna(how='all'): drop the rows where all of the elements are missing.
* df.dropna(subset=['PassengerId']): define in which columns to look for missing values.

If we want to drop the rows with at least one missing element:

In [ ]:
df.dropna().shape

(183, 12)

If we want to drop the rows with all elements missing:

In [ ]:
df.dropna(how='all')

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

If we want to drop the rows that are missing PassengerID.

In [ ]:
df.dropna(subset=['PassengerId']).shape

(891, 12)

If we want to update the dataset after dropping rows, we can use inplace = True

In [ ]:
df.dropna(subset=['PassengerId'], inplace=True)

In the titanic dataset, we've confirmed that all of the rows have 'PassengerId' because the number of rows is the same as the code above.

### Now Try This

Drop the rows that are missing any of the following columns: 'PassengerId', 'Survived', 'Age'

In [ ]:
df.dropna(subset=['PassengerId', 'Survived', 'Age'])

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
885          886         0       3   
886          887         0       2   
887          888         1       1   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
885      5            382652  29.1250   NaN        Q  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[714 rows x 12 columns]

## Removing a column

Before we dive into the data analysis, let's see if there are any columns we want to remove. 

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 90.5+ KB


In "Cabin" column, there are only 204 rows that are non-null. That means 891 - 204 = 787 rows are missing in the column.

So, it wouldn't give us a meaningful insight and we can remove "Cabin" column by using ```del``` function.

In [ ]:
del df['Cabin']
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Embarked  
0        0         A/5 21171   7.2500        S  
1        0          PC 17599  71.2833        C  
2        0  STON/O2. 3101282   7.9250        S  
3        0            113803  53.1000        S  
4        0            373450   8.0500        S  
..     ...               ...      ...      ...  
886      0            211536  13.0000        S  
887      0            112053  30.0000        S  
888      2        W./C. 6607  23.4500        S  
889      0            111369  30.0000        C  
890      0            370376   7.7500        Q  

[891 rows x 11 columns]

## Select subsets of data

When we are interested in a few columns to do the data analysis, we can select a specific subset of columns using two methods:

1. by index location
2. by column names

We can select specific subsets of data using *iloc[rows_index, columns_index]*.

As we learned last week, [:] selects everything in a list or string in Python. 

Similarly, [:] will select every row or column depending on where we put it.

In [ ]:
# by index location (iloc)
df_index = df.iloc[: , [0,1,2,3]]
df_index

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name  
0                              Braund, Mr. Owen Harris  
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  
2                               Heikkinen, Miss. Laina  
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  
4                             Allen, Mr. William Henry  
..                                                 ...  
886                              Montvila, Rev. Juozas  
887                       Graham, Miss. Margaret Edith  
888           Johnston, Miss. Catherine Helen "Carrie"  
889                              Behr, Mr. Karl Howell  
890                                Dooley, Mr. Patrick  

[891 rows x 4 columns]

In [ ]:
# by column names
df_col_names = df[['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age']]
df_col_names

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  
0                              Braund, Mr. Owen Harris    male  22.0  
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0  
2                               Heikkinen, Miss. Laina  female  26.0  
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0  
4                             Allen, Mr. William Henry    male  35.0  
..                                                 ...     ...   ...  
886                              Montvila, Rev. Juozas    male  27.0  
887                       Graham, Miss. Margaret Edith  female  19.0  
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN  
889                              Behr, Mr. Karl Howell    male  26.0  
890                                Dooley, Mr. Patrick    male  32.0  

[891 rows x 6 columns]

### Now Try This

Select column index 3 to column index 6 (inclusive) with all rows.

In [ ]:
df_index = df.iloc[: , 3:7]
df_index

Name     Sex   Age  SibSp
0                              Braund, Mr. Owen Harris    male  22.0      1
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1
2                               Heikkinen, Miss. Laina  female  26.0      0
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1
4                             Allen, Mr. William Henry    male  35.0      0
..                                                 ...     ...   ...    ...
886                              Montvila, Rev. Juozas    male  27.0      0
887                       Graham, Miss. Margaret Edith  female  19.0      0
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1
889                              Behr, Mr. Karl Howell    male  26.0      0
890                                Dooley, Mr. Patrick    male  32.0      0

[891 rows x 4 columns]

## Filter Dataset based on criteria

Often times, we are interested in working with specific rows that meet the certain criteria. 

If we only want to look at the data with Age > 30, we can specify the criteria within the 'loc' function.

In [ ]:
df_over_30yrs=df.loc[df['Age'] > 30]
df_over_30yrs

PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
4              5         0       3   
6              7         0       1   
11            12         1       1   
..           ...       ...     ...   
873          874         0       3   
879          880         1       1   
881          882         0       3   
885          886         0       3   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
..                                                 ...     ...   ...    ...   
873                        Vander Cruyssen, Mr. Victor    male  47.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
881                                 Markun, Mr. Johann    male  33.0      0   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch    Ticket     Fare Embarked  
1        0  PC 17599  71.2833        C  
3        0    113803  53.1000        S  
4        0    373450   8.0500        S  
6        0     17463  51.8625        S  
11       0    113783  26.5500        S  
..     ...       ...      ...      ...  
873      0    345765   9.0000        S  
879      1     11767  83.1583        C  
881      0    349257   7.8958        S  
885      5    382652  29.1250        Q  
890      0    370376   7.7500        Q  

[305 rows x 11 columns]

Now, let's select the dataset using two criteria -- where "Age" is greater than 30 AND "Survived."

'*&*' is equivalent to '*AND*' and '*|*' is equivalent to '*OR*' in dataframe.


In [ ]:
df_over_30yrs_survived = df.loc[(df['Age'] > 30) & (df['Survived'] == 1)]
df_over_30yrs_survived

PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
11            12         1       1   
15            16         1       2   
21            22         1       2   
..           ...       ...     ...   
857          858         1       1   
862          863         1       1   
865          866         1       2   
871          872         1       1   
879          880         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
21                               Beesley, Mr. Lawrence    male  34.0      0   
..                                                 ...     ...   ...    ...   
857                             Daly, Mr. Peter Denis     male  51.0      0   
862  Swift, Mrs. Frederick Joel (Margaret Welles Ba...  female  48.0      0   
865                           Bystrom, Mrs. (Karolina)  female  42.0      0   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   

     Parch    Ticket     Fare Embarked  
1        0  PC 17599  71.2833        C  
3        0    113803  53.1000        S  
11       0    113783  26.5500        S  
15       0    248706  16.0000        S  
21       0    248698  13.0000        S  
..     ...       ...      ...      ...  
857      0    113055  26.5500        S  
862      0     17466  25.9292        S  
865      0    236852  13.0000        S  
871      1     11751  52.5542        S  
879      1     11767  83.1583        C  

[124 rows x 11 columns]


IMPORTANT: When filtering with multiple conditions, make sure to use **()** on each condition.

Let's check how many passengers survived among the ones whose age was over 30.

In [ ]:
print("# of passengers whose age was over 30: ", df_over_30yrs.shape[0])
print("# of survived passengers whose age was over 30: ", df_over_30yrs_survived.shape[0])

### Now Try This

Select the dataset that meet the following condition:
- Pclass is not 1

In [ ]:
df.loc[(df['Pclass'] != 1)]

PassengerId  Survived  Pclass                                      Name  \
0              1         0       3                   Braund, Mr. Owen Harris   
2              3         1       3                    Heikkinen, Miss. Laina   
4              5         0       3                  Allen, Mr. William Henry   
5              6         0       3                          Moran, Mr. James   
7              8         0       3            Palsson, Master. Gosta Leonard   
..           ...       ...     ...                                       ...   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Embarked  
0      male  22.0      1      0         A/5 21171   7.2500        S  
2    female  26.0      0      0  STON/O2. 3101282   7.9250        S  
4      male  35.0      0      0            373450   8.0500        S  
5      male   NaN      0      0            330877   8.4583        Q  
7      male   2.0      3      1            349909  21.0750        S  
..      ...   ...    ...    ...               ...      ...      ...  
884    male  25.0      0      0   SOTON/OQ 392076   7.0500        S  
885  female  39.0      0      5            382652  29.1250        Q  
886    male  27.0      0      0            211536  13.0000        S  
888  female   NaN      1      2        W./C. 6607  23.4500        S  
890    male  32.0      0      0            370376   7.7500        Q  

[675 rows x 11 columns]

### Now Try This

Select the dataset that meet the following conditions:
- Age is less than 10 
- OR
- Age is greater than 50

Hint: Don't forget parenthesis!

In [ ]:
df.loc[(df['Age'] < 10) | (df['Age'] > 50)]

PassengerId  Survived  Pclass  \
6              7         0       1   
7              8         0       3   
10            11         1       3   
11            12         1       1   
15            16         1       2   
..           ...       ...     ...   
851          852         0       3   
852          853         0       3   
857          858         1       1   
869          870         1       3   
879          880         1       1   

                                              Name     Sex   Age  SibSp  \
6                          McCarthy, Mr. Timothy J    male  54.0      0   
7                   Palsson, Master. Gosta Leonard    male   2.0      3   
10                 Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                        Bonnell, Miss. Elizabeth  female  58.0      0   
15                Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
..                                             ...     ...   ...    ...   
851                            Svensson, Mr. Johan    male  74.0      0   
852                        Boulos, Miss. Nourelain  female   9.0      1   
857                         Daly, Mr. Peter Denis     male  51.0      0   
869                Johnson, Master. Harold Theodor    male   4.0      1   
879  Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   

     Parch   Ticket     Fare Embarked  
6        0    17463  51.8625        S  
7        1   349909  21.0750        S  
10       1  PP 9549  16.7000        S  
11       0   113783  26.5500        S  
15       0   248706  16.0000        S  
..     ...      ...      ...      ...  
851      0   347060   7.7750        S  
852      1     2678  15.2458        C  
857      0   113055  26.5500        S  
869      1   347742  11.1333        S  
879      1    11767  83.1583        C  

[126 rows x 11 columns]

## Aggregation

Aggregation is the process of combining things. It's useful to understand overall properties of the dataset and analyze it.

Some examples of aggregation are sum(), min(), max(), count(), mean(), std(), etc.

### Sum
#### 1. Total fares

In [ ]:
df['Fare'].sum()

28693.9493

If we want to round total fares and save it as a variable, then we can try:

In [ ]:
total_fares = df['Fare'].sum()
print("Total fares: ", round(total_fares))

Total fares:  28694.0


#### 2. Survived passengers
We can also count the number of passengers that survived by summing up the 'Survived' column.

In [ ]:
survived_passengers = df['Survived'].sum()
survived_passengers

342

### Max / Min

Let's calculate max and min of Fare.

In [ ]:
df['Fare'].max()

512.3292

In [ ]:
df['Fare'].min()

0.0

### Mean
Let's calculate the survival rate of all passengers.

In [ ]:
df['Survived'].mean()

0.3838383838383838

Let's calculate the average age of all passengers.

In [ ]:
df['Age'].mean()

29.69911764705882

Now, we will tackle a more complex problem.

Let's calculate the survival rate by the *age* group.
We can apply filtering that we just learned to select the group whose age was over 30 and whose age was under 30.

In each group, we will calculate the mean of 'Survived' column.

In [ ]:
# filtering
df_over_30yrs = df.loc[df['Age'] > 30]
df_under_30yrs = df.loc[df['Age'] <= 30]

# calculating mean of 'Survived' for each group
mean_over_30 = df_over_30yrs['Survived'].mean()
mean_under_30 = df_under_30yrs['Survived'].mean()

# printing the mean survival rates for each group
# round(number, decimal_points): round a number to a given precision in decimal_points
print("Survival rate - age over 30: ", round(mean_over_30*100, 3), "%")
print("Survival rate - age under 30: ", round(mean_under_30*100, 3), "%")

Survival rate - age over 30:  40.656 %
Survival rate - age under 30:  40.587 %


There's not much difference between the two groups.

### Groupby

Now, we will group by *sex* to see if there's any difference between female and male.

Here, we use 'groupby' aggregate function and it will let us group the dataset by that column ('Sex')

In [ ]:
df.groupby(['Sex']).mean()

PassengerId  Survived    Pclass        Age     SibSp     Parch  \
Sex                                                                      
female   431.028662  0.742038  2.159236  27.915709  0.694268  0.649682   
male     454.147314  0.188908  2.389948  30.726645  0.429809  0.235702   

             Fare  
Sex                
female  44.479818  
male    25.523893

If we want to sort the aggregate funtion by a column, we can use sort_values(by="column_name").

Let's sort the above aggregate function by "Survived"

In [ ]:
df.groupby(['Sex']).mean().sort_values(by="Survived")

PassengerId  Survived    Pclass        Age     SibSp     Parch  \
Sex                                                                      
male     454.147314  0.188908  2.389948  30.726645  0.429809  0.235702   
female   431.028662  0.742038  2.159236  27.915709  0.694268  0.649682   

             Fare  
Sex                
male    25.523893  
female  44.479818

If we are interested in the survival rate of each group and we specify it by using [ ] after the groupby call.

In [ ]:
# group by 'Sex' and calculate mean of 'Survived'
df.groupby(['Sex'])['Survived'].mean()

Sex
female    0.742038
male      0.188908
Name: Survived, dtype: float64

There was a significant difference in the survival rate by *sex*!

We can also apply groupby on multiple columns.

In [ ]:
 df.groupby(['Sex', 'Pclass']).mean()

PassengerId  Survived        Age     SibSp     Parch  \
Sex    Pclass                                                         
female 1        469.212766  0.968085  34.611765  0.553191  0.457447   
       2        443.105263  0.921053  28.722973  0.486842  0.605263   
       3        399.729167  0.500000  21.750000  0.895833  0.798611   
male   1        455.729508  0.368852  41.281386  0.311475  0.278689   
       2        447.962963  0.157407  30.740707  0.342593  0.222222   
       3        455.515850  0.135447  26.507589  0.498559  0.224784   

                     Fare  
Sex    Pclass              
female 1       106.125798  
       2        21.970121  
       3        16.118810  
male   1        67.226127  
       2        19.741782  
       3        12.661633

In both groups (female and male), the survival rate was a lot higher in Pclass 1 than other Pclass!

### Now Try This

Then, some of us might be curious:

Would lower Pclass be more expensive or higher Pclass be more expensive?

We can answer the question by calculating the mean fares for each class.

**Calculate the mean fares for each Pclass!**

In [ ]:
df.groupby(['Pclass'])['Fare'].mean()

Pclass
1    84.154687
2    20.662183
3    13.675550
Name: Fare, dtype: float64

### Let's pause and think!
Did you see any correlation between Pclass, Fare, and Survival rate? Briefly describe what you have found here.

## Takeaways

Using Dataframe and aggregate functions in Pandas, we can answer any questions that might come up!

In the tutorial section, we will apply what we have learned in Pandas and further analyze a new dataset.

# Tutorial

### About the dataset

What we will be using in the tutorial is US Census Demographic Data.

The data here were collected by the US Census Burea and it includes data from the entire country.

This dataset covers lots of areas: state, county, gender, ethnicity, working fields, commuting modes, and employment.

All of this information is available at a State and County level. There are many questions that we could try to answer with the dataset:
- Unemployment by state
- Professional fields by state and county
- Commute transit mode by county in CA
- ...

### Objective
Since the dataset covers all of the states in the US, we are going to select top 5 states by population. 
Once we've selected top five states, we will calculate the average income and the total employed population.

That's our focus in the tutorial, but feel free to play around with it as you'd like.

Let's load our data first!

### Load file

In [10]:
df = pd.read_csv('acs2017_county_data.csv')
df

CountyId        State  ... FamilyWork  Unemployment
0         1001      Alabama  ...        0.1           5.2
1         1003      Alabama  ...        0.1           5.5
2         1005      Alabama  ...        0.3          12.4
3         1007      Alabama  ...        0.3           8.2
4         1009      Alabama  ...        0.1           4.9
...        ...          ...  ...        ...           ...
3215     72145  Puerto Rico  ...        0.2          16.8
3216     72147  Puerto Rico  ...        0.0          12.8
3217     72149  Puerto Rico  ...        0.2          24.8
3218     72151  Puerto Rico  ...        0.0          25.4
3219     72153  Puerto Rico  ...        0.0          24.0

[3220 rows x 37 columns]

### Basic info about the dataset


In [ ]:
df.head()

CountyId    State          County  TotalPop    Men   Women  Hispanic  \
0      1001  Alabama  Autauga County     55036  26899   28137       2.7   
1      1003  Alabama  Baldwin County    203360  99527  103833       4.4   
2      1005  Alabama  Barbour County     26201  13976   12225       4.2   
3      1007  Alabama     Bibb County     22580  12251   10329       2.4   
4      1009  Alabama   Blount County     57667  28490   29177       9.0   

   White  Black  Native  ...  Walk  OtherTransp  WorkAtHome  MeanCommute  \
0   75.4   18.9     0.3  ...   0.6          1.3         2.5         25.8   
1   83.1    9.5     0.8  ...   0.8          1.1         5.6         27.0   
2   45.7   47.8     0.2  ...   2.2          1.7         1.3         23.4   
3   74.6   22.0     0.4  ...   0.3          1.7         1.5         30.0   
4   87.4    1.5     0.3  ...   0.4          0.4         2.1         35.0   

   Employed  PrivateWork  PublicWork  SelfEmployed  FamilyWork  Unemployment  
0     24112         74.1        20.2           5.6         0.1           5.2  
1     89527         80.7        12.9           6.3         0.1           5.5  
2      8878         74.1        19.1           6.5         0.3          12.4  
3      8171         76.0        17.4           6.3         0.3           8.2  
4     21380         83.9        11.9           4.0         0.1           4.9  

[5 rows x 37 columns]

Since there are so many columns, head function doesn't display all columns.

Let's use info() as it returns **ALL** of the **column names**, its types, and Non-Null counts. 



In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3220 entries, 0 to 3219
Data columns (total 37 columns):
CountyId            3220 non-null int64
State               3220 non-null object
County              3220 non-null object
TotalPop            3220 non-null int64
Men                 3220 non-null int64
Women               3220 non-null int64
Hispanic            3220 non-null float64
White               3220 non-null float64
Black               3220 non-null float64
Native              3220 non-null float64
Asian               3220 non-null float64
Pacific             3220 non-null float64
VotingAgeCitizen    3220 non-null int64
Income              3220 non-null int64
IncomeErr           3220 non-null int64
IncomePerCap        3220 non-null int64
IncomePerCapErr     3220 non-null int64
Poverty             3220 non-null float64
ChildPoverty        3219 non-null float64
Professional        3220 non-null float64
Service             3220 non-null float64
Office              3220 non-nu

There are 37 columns with 3220 counties.

Also, none of the columns have missing rows as every column has 3220 non-null values!

Let's view summary statistics of numeric columns and figure out what we want to get out from this dataset.

In [ ]:
df.describe()

CountyId      TotalPop           Men         Women     Hispanic  \
count   3220.000000  3.220000e+03  3.220000e+03  3.220000e+03  3220.000000   
mean   31393.605280  1.007681e+05  4.958781e+04  5.118032e+04    11.296584   
std    16292.078954  3.244996e+05  1.593212e+05  1.652164e+05    19.342522   
min     1001.000000  7.400000e+01  3.900000e+01  3.500000e+01     0.000000   
25%    19032.500000  1.121350e+04  5.645500e+03  5.553500e+03     2.100000   
50%    30024.000000  2.584750e+04  1.287900e+04  1.299350e+04     4.100000   
75%    46105.500000  6.660825e+04  3.301725e+04  3.359375e+04    10.000000   
max    72153.000000  1.010572e+07  4.979641e+06  5.126081e+06   100.000000   

             White        Black       Native        Asian      Pacific  ...  \
count  3220.000000  3220.000000  3220.000000  3220.000000  3220.000000  ...   
mean     74.920186     8.681957     1.768416     1.289379     0.083416  ...   
std      23.056700    14.333571     7.422946     2.716191     0.709277  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%      63.500000     0.600000     0.100000     0.200000     0.000000  ...   
50%      83.600000     2.000000     0.300000     0.600000     0.000000  ...   
75%      92.800000     9.500000     0.600000     1.200000     0.100000  ...   
max     100.000000    86.900000    90.300000    41.800000    33.700000  ...   

              Walk  OtherTransp   WorkAtHome  MeanCommute      Employed  \
count  3220.000000  3220.000000  3220.000000  3220.000000  3.220000e+03   
mean      3.244472     1.598696     4.736894    23.474534  4.709295e+04   
std       3.891510     1.678232     3.073484     5.687241  1.558159e+05   
min       0.000000     0.000000     0.000000     5.100000  3.900000e+01   
25%       1.400000     0.800000     2.900000    19.600000  4.573000e+03   
50%       2.300000     1.300000     4.100000    23.200000  1.061150e+04   
75%       3.825000     1.900000     5.800000    27.000000  2.874725e+04   
max      59.200000    43.200000    33.000000    45.100000  4.805817e+06   

       PrivateWork   PublicWork  SelfEmployed   FamilyWork  Unemployment  
count  3220.000000  3220.000000   3220.000000  3220.000000   3220.000000  
mean     74.863323    17.086118      7.772733     0.278820      6.665590  
std       7.647916     6.390868      3.855454     0.448073      3.772612  
min      31.100000     4.400000      0.000000     0.000000      0.000000  
25%      71.200000    12.700000      5.200000     0.100000      4.475000  
50%      76.100000    15.900000      6.800000     0.200000      6.100000  
75%      80.200000    19.900000      9.200000     0.300000      8.000000  
max      88.800000    64.800000     38.000000     8.000000     40.900000  

[8 rows x 35 columns]

It says that this dataset 35 numerical columns, but actually CountyId is not a numerical value. It's a unique identifier for each county.

As we saw in the objective, there are lots of things we can do with this dataset.

But, let's pick top 5 states with most population and most employees and work from there!


## Select subsets of data

Before we dive into the data analysis, let's select the columns we want to work with. 
As we learned earlier, we can select a specific subset of columns using two methods:

1. by index location
2. by column names

As this dataset has so many columns, let's take a look at all of the columns first.

We can call columns and it will return all of the column names. (or we can use df.info() as well.)

In [ ]:
df.columns

Index(['CountyId', 'State', 'County', 'TotalPop', 'Men', 'Women', 'Hispanic',
       'White', 'Black', 'Native', 'Asian', 'Pacific', 'VotingAgeCitizen',
       'Income', 'IncomeErr', 'IncomePerCap', 'IncomePerCapErr', 'Poverty',
       'ChildPoverty', 'Professional', 'Service', 'Office', 'Construction',
       'Production', 'Drive', 'Carpool', 'Transit', 'Walk', 'OtherTransp',
       'WorkAtHome', 'MeanCommute', 'Employed', 'PrivateWork', 'PublicWork',
       'SelfEmployed', 'FamilyWork', 'Unemployment'],
      dtype='object')

Since there are so many columns, it's hard to count the indices. So we will use the column names to select subsets of data!

In [ ]:
# by column names
df_emp = df[['CountyId', 'State', 'County', 'TotalPop', 'Income', 'Employed', 'Drive', 'Carpool', 'Transit', 'Walk', 'OtherTransp', 'WorkAtHome']]
df_emp

CountyId        State               County  TotalPop  Income  Employed  \
0         1001      Alabama       Autauga County     55036   55317     24112   
1         1003      Alabama       Baldwin County    203360   52562     89527   
2         1005      Alabama       Barbour County     26201   33368      8878   
3         1007      Alabama          Bibb County     22580   43404      8171   
4         1009      Alabama        Blount County     57667   47412     21380   
...        ...          ...                  ...       ...     ...       ...   
3215     72145  Puerto Rico  Vega Baja Municipio     54754   18900     14234   
3216     72147  Puerto Rico    Vieques Municipio      8931   16261      2927   
3217     72149  Puerto Rico   Villalba Municipio     23659   19893      6873   
3218     72151  Puerto Rico    Yabucoa Municipio     35025   15586      7878   
3219     72153  Puerto Rico      Yauco Municipio     37585   14451      8995   

      Drive  Carpool  Transit  Walk  OtherTransp  WorkAtHome  
0      86.0      9.6      0.1   0.6          1.3         2.5  
1      84.7      7.6      0.1   0.8          1.1         5.6  
2      83.4     11.1      0.3   2.2          1.7         1.3  
3      86.4      9.5      0.7   0.3          1.7         1.5  
4      86.8     10.2      0.1   0.4          0.4         2.1  
...     ...      ...      ...   ...          ...         ...  
3215   92.0      4.2      0.9   1.4          0.6         0.9  
3216   76.3     16.9      0.0   5.0          0.0         1.7  
3217   83.1     11.8      0.1   2.1          0.0         2.8  
3218   87.6      9.2      0.0   1.4          1.8         0.1  
3219   82.8      8.2      2.2   1.7          0.1         5.0  

[3220 rows x 12 columns]

## Aggregation

### Total population by state

In [ ]:
state_pop = df_emp.groupby(['State'])['TotalPop'].sum()
state_pop

State
Alabama                  4850771
Alaska                    738565
Arizona                  6809946
Arkansas                 2977944
California              38982847
Colorado                 5436519
Connecticut              3594478
Delaware                  943732
District of Columbia      672391
Florida                 20278447
Georgia                 10201635
Hawaii                   1421658
Idaho                    1657375
Illinois                12854526
Indiana                  6614418
Iowa                     3118102
Kansas                   2903820
Kentucky                 4424376
Louisiana                4663461
Maine                    1330158
Maryland                 5996079
Massachusetts            6789319
Michigan                 9925568
Minnesota                5490726
Mississippi              2986220
Missouri                 6075300
Montana                  1029862
Nebraska                 1893921
Nevada                   2887725
New Hampshire            1331848
New 

## Total population by state -- sorted

Since there are so many states, it will be easier to see which states have most population if we can sort the dataset.

If we use sort_values() it's going to sort the data by the aggregation function (sum)

In [ ]:
state_pop = df_emp.groupby(['State'])['TotalPop'].sum().sort_values(ascending=False)
state_pop

State
California              38982847
Texas                   27419612
Florida                 20278447
New York                19798228
Illinois                12854526
Pennsylvania            12790505
Ohio                    11609756
Georgia                 10201635
North Carolina          10052564
Michigan                 9925568
New Jersey               8960161
Virginia                 8365952
Washington               7169967
Arizona                  6809946
Massachusetts            6789319
Indiana                  6614418
Tennessee                6597381
Missouri                 6075300
Maryland                 5996079
Wisconsin                5763217
Minnesota                5490726
Colorado                 5436519
South Carolina           4893444
Alabama                  4850771
Louisiana                4663461
Kentucky                 4424376
Oregon                   4025127
Oklahoma                 3896251
Connecticut              3594478
Puerto Rico              3468963
Iowa

Let's select top FIVE states with most population!

#### These are the five states we will work with:
- California             
- Texas                   
- Florida                 
- New York                
- Illinois  

We are going to use "loc" as we want to filter dataset based on criteria.

We could use the line below: selecting rows if 'State' is California, Texas, Florida, New York, or Illinois.

In [ ]:
df_emp.loc[(df_emp['State']=='California') | (df_emp['State']=='Texas') | (df_emp['State']=='Florida') | (df_emp['State']== "New York") | (df_emp['State']=='Illinois')]

CountyId       State            County  TotalPop  Income  Employed  \
186       6001  California    Alameda County   1629615   85743    826310   
187       6003  California     Alpine County      1203   63438       374   
188       6005  California     Amador County     37306   60636     13444   
189       6007  California      Butte County    225207   46516     93439   
190       6009  California  Calaveras County     45057   54800     16721   
...        ...         ...               ...       ...     ...       ...   
2772     48499       Texas       Wood County     43315   48038     15960   
2773     48501       Texas     Yoakum County      8481   62500      3755   
2774     48503       Texas      Young County     18166   46351      8248   
2775     48505       Texas     Zapata County     14415   34550      5146   
2776     48507       Texas     Zavala County     12152   25988      4397   

      Drive  Carpool  Transit  Walk  OtherTransp  WorkAtHome  
186    62.0     10.1     14.6   3.7          3.6         6.0  
187    62.1     13.2      0.0   9.1          1.6        14.0  
188    80.7      9.2      0.1   2.5          1.1         6.4  
189    75.0     11.1      1.2   3.2          3.9         5.7  
190    79.1      9.9      1.0   1.2          0.6         8.3  
...     ...      ...      ...   ...          ...         ...  
2772   80.6     10.0      0.3   2.5          1.0         5.6  
2773   78.0     16.8      0.0   3.7          0.2         1.3  
2774   83.0      9.0      0.0   0.6          1.4         6.1  
2775   69.2     23.8      0.0   4.4          0.6         2.0  
2776   80.5      7.4      0.4   2.1          4.2         5.4  

[543 rows x 12 columns]

But, the code above is very lenthy so we will learn a shortcut!

We can use isin(state_list) function to see if 'State' is in state_list.

The syntax above is very similar to 'a' in ['a','b','c'] in Python.

In [ ]:
five_states = df_emp.loc[df_emp['State'].isin(['California','Texas','Florida','New York','Illinois'])]
five_states

CountyId       State            County  TotalPop  Income  Employed  \
186       6001  California    Alameda County   1629615   85743    826310   
187       6003  California     Alpine County      1203   63438       374   
188       6005  California     Amador County     37306   60636     13444   
189       6007  California      Butte County    225207   46516     93439   
190       6009  California  Calaveras County     45057   54800     16721   
...        ...         ...               ...       ...     ...       ...   
2772     48499       Texas       Wood County     43315   48038     15960   
2773     48501       Texas     Yoakum County      8481   62500      3755   
2774     48503       Texas      Young County     18166   46351      8248   
2775     48505       Texas     Zapata County     14415   34550      5146   
2776     48507       Texas     Zavala County     12152   25988      4397   

      Drive  Carpool  Transit  Walk  OtherTransp  WorkAtHome  
186    62.0     10.1     14.6   3.7          3.6         6.0  
187    62.1     13.2      0.0   9.1          1.6        14.0  
188    80.7      9.2      0.1   2.5          1.1         6.4  
189    75.0     11.1      1.2   3.2          3.9         5.7  
190    79.1      9.9      1.0   1.2          0.6         8.3  
...     ...      ...      ...   ...          ...         ...  
2772   80.6     10.0      0.3   2.5          1.0         5.6  
2773   78.0     16.8      0.0   3.7          0.2         1.3  
2774   83.0      9.0      0.0   0.6          1.4         6.1  
2775   69.2     23.8      0.0   4.4          0.6         2.0  
2776   80.5      7.4      0.4   2.1          4.2         5.4  

[543 rows x 12 columns]

### Average income by state

Now, we have selected five states to work with, and let's get the average income in each state.

In [ ]:
five_states.groupby(['State'])['Income'].mean().sort_values(ascending=False)

State
California    61046.758621
New York      58309.258065
Illinois      52930.382353
Texas         49894.338583
Florida       47144.328358
Name: Income, dtype: float64

California has the highest average income and Florida has the lowest average income amongst five states.

### Total number of employees by state
### Now Try This
Now, let's look at the employed population by each state.

In [ ]:
five_states.groupby(['State'])['Employed'].sum().sort_values(ascending=False)

State
California    17993915
Texas         12689069
New York       9467631
Florida        9018570
Illinois       6181653
Name: Employed, dtype: int64

Write down the two states with the highest employed population and the lowest employed population.



### Transit modes by state

Let's look at each state's transit mode and see which transit mode is most popular in each state.

We will groupby 'State' and we will get the sum of all of the transit modes.

In [ ]:
five_states.groupby(['State'])['Drive','Carpool','Transit','Walk','OtherTransp','WorkAtHome'].sum()

Drive  Carpool  Transit   Walk  OtherTransp  WorkAtHome
State                                                                
California   4265.3    633.3    170.1  193.5        151.0       386.8
Florida      5349.1    687.6     72.2   98.2        157.0       335.6
Illinois     8420.2    872.8    108.7  248.1        126.7       422.4
New York     4619.3    523.4    406.5  290.6        100.0       260.2
Texas       20455.0   2888.5     79.4  600.6        363.2      1013.9

### Now Try This

Step 1: Select california state only

Step 2: Calculate the sum of all transit modes by county in California.

In [ ]:
# step 1: Select california state only
ca_transit = five_states.loc[five_states['State']=='California']
ca_transit

CountyId       State                  County  TotalPop  Income  Employed  \
186      6001  California          Alameda County   1629615   85743    826310   
187      6003  California           Alpine County      1203   63438       374   
188      6005  California           Amador County     37306   60636     13444   
189      6007  California            Butte County    225207   46516     93439   
190      6009  California        Calaveras County     45057   54800     16721   
191      6011  California           Colusa County     21479   56481      9470   
192      6013  California     Contra Costa County   1123678   88456    535401   
193      6015  California        Del Norte County     27442   41287      8742   
194      6017  California        El Dorado County    185015   74885     80373   
195      6019  California           Fresno County    971616   48730    392003   
196      6021  California            Glenn County     27935   46260     10775   
197      6023  California         Humboldt County    135490   43718     59592   
198      6025  California         Imperial County    179957   44779     59757   
199      6027  California             Inyo County     18195   51500      8275   
200      6029  California             Kern County    878744   50826    335688   
201      6031  California            Kings County    150183   49742     53158   
202      6033  California             Lake County     64095   40446     22691   
203      6035  California           Lassen County     31470   54083      9052   
204      6037  California      Los Angeles County  10105722   61015   4805817   
205      6039  California           Madera County    154440   48210     56076   
206      6041  California            Marin County    260814  104703    130570   
207      6043  California         Mariposa County     17658   51385      7009   
208      6045  California        Mendocino County     87497   46528     37084   
209      6047  California           Merced County    267390   46338     99026   
210      6049  California            Modoc County      9017   39296      3231   
211      6051  California             Mono County     14058   60595      7864   
212      6053  California         Monterey County    433168   63249    182439   
213      6055  California             Napa County    141005   79637     71110   
214      6057  California           Nevada County     98838   60610     42302   
215      6059  California           Orange County   3155816   81851   1560997   
216      6061  California           Placer County    374985   80488    170791   
217      6063  California           Plumas County     18724   50266      7260   
218      6065  California        Riverside County   2355002   60807    978726   
219      6067  California       Sacramento County   1495400   60239    666790   
220      6069  California       San Benito County     58671   80760     27688   
221      6071  California   San Bernardino County   2121220   57156    869658   
222      6073  California        San Diego County   3283665   70588   1536073   
223      6075  California    San Francisco County    864263   96265    504914   
224      6077  California      San Joaquin County    724153   57813    296264   
225      6079  California  San Luis Obispo County    280119   67175    129277   
226      6081  California        San Mateo County    763450  105667    404882   
227      6083  California    Santa Barbara County    442996   68023    211793   
228      6085  California      Santa Clara County   1911226  106761    967294   
229      6087  California       Santa Cruz County    273263   73663    133847   
230      6089  California           Shasta County    178919   47258     72195   
231      6091  California           Sierra County      2885   44190      1149   
232      6093  California         Siskiyou County     43530   40884     16544   
233      6095  California           Solano County    434981   72950    196309   
234      6097  California           

In [ ]:
# Step 2: Calculate the sum of all transit modes by county in California.

ca_transit.groupby(['County'])['Drive','Carpool','Transit','Walk','OtherTransp','WorkAtHome'].sum()

Drive  Carpool  Transit  Walk  OtherTransp  WorkAtHome
County                                                                        
Alameda County           62.0     10.1     14.6   3.7          3.6         6.0
Alpine County            62.1     13.2      0.0   9.1          1.6        14.0
Amador County            80.7      9.2      0.1   2.5          1.1         6.4
Butte County             75.0     11.1      1.2   3.2          3.9         5.7
Calaveras County         79.1      9.9      1.0   1.2          0.6         8.3
Colusa County            76.9     14.2      0.3   2.5          1.6         4.5
Contra Costa County      68.1     11.7     10.3   1.7          1.9         6.3
Del Norte County         73.7     14.5      0.6   3.6          1.9         5.6
El Dorado County         77.8      8.5      1.8   2.0          1.6         8.3
Fresno County            78.1     12.2      1.2   1.6          2.5         4.3
Glenn County             75.9     12.4      0.4   3.9          1.6         5.7
Humboldt County          72.0      9.1      1.9   6.7          2.8         7.4
Imperial County          81.4      8.8      0.8   1.9          2.6         4.4
Inyo County              69.4     12.5      1.3   7.2          5.2         4.5
Kern County              79.2     12.8      0.9   1.2          2.7         3.2
Kings County             75.6     14.4      0.9   2.2          2.0         5.0
Lake County              71.3     11.5      1.0   3.0          0.6        12.5
Lassen County            80.1      9.8      0.7   2.8          0.9         5.8
Los Angeles County       73.7      9.6      6.3   2.7          2.3         5.3
Madera County            76.4     15.5      0.5   2.0          2.0         3.6
Marin County             65.3      8.1      9.8   3.7          2.4        10.7
Mariposa County          62.1     12.9      2.4   9.5          3.9         9.2
Mendocino County         74.6     10.1      0.3   5.2          1.7         8.1
Merced County            78.8     10.8      1.2   2.1          3.7         3.4
Modoc County             70.3      7.4      0.5   5.9          0.7        15.1
Mono County              44.8      7.0     19.7  11.1          8.1         9.3
Monterey County          71.1     10.7      1.7   3.1          8.8         4.6
Napa County              76.1     11.5      1.4   3.9          1.6         5.6
Nevada County            74.4      8.6      1.2   2.8          2.7        10.4
Orange County            78.6      9.8      2.2   1.9          2.0         5.6
Placer County            79.5      7.8      1.3   1.5          1.6         8.3
Plumas County            78.5     10.3      0.1   3.1          2.3         5.7
Riverside County         77.2     12.9      1.3   1.6          1.7         5.2
Sacramento County        77.1     10.4      2.7   1.9          2.4         5.5
San Benito County        80.3     13.5      0.5   1.2          1.1         3.4
San Bernardino County    78.9     12.0      1.5   1.7          1.2         4.7
San Diego County         76.0      8.9      3.1   2.9          2.2         7.0
San Francisco County     34.3      6.8     34.0  11.1          7.1         6.7
San Joaquin County       77.4     13.9      1.5   1.6          1.5         4.0
San Luis Obispo County   74.8      9.9      1.4   3.9          3.2         6.7
San Mateo County         68.7     10.6     10.4   2.6          2.8         4.8
Santa Barbara County     68.0     13.6      3.2   4.6          4.8         5.8
Santa Clara County       75.1     10.5      4.4   2.1          3.1         4.8
Santa Cruz County        69.2      9.0      2.8   4.5          7.0         7.5
Shasta County            82.0      7.5      0.6   2.2          1.9         5.8
Sierra County            66.5     11.4      0.0   3.4          0.7        18.1
Siskiyou County          74.4      9.9      0.7   4.9          1.3         8.9
Solano County            76.6     13.6      3.0   1.3          1.4         4.1
Sonoma County            75.0     11.0      1.9   3.1          2.1        

## Takeaways from this tutorial
Anytime we want to work with data, we can use Pandas, and we will be able to answer any questions!

In the later weeks, we will deep dive into the further uses of Pandas to analyze a more complicated data.

## Resources
- [10 minutes to Pandas](https://pandas.pydata.org/pandas-docs/stable/user_guide/10min.html)
- [Pandas API Reference](https://pandas.pydata.org/pandas-docs/stable/reference/index.html)
- [A Gentle Introduction to Pandas](https://medium.com/@wbusaka/a-gentle-introduction-to-pandas-5ed17421a59d)

## Homework

## Appendix